content/Testing in Django - Basics - useful tools and good practices.md

```diff
@@ -0,0 +1,352 @@
+- You can define an attribute in a factory that selects a choice of a set of choices just like a Django model choice field, using `random_element` of Faker. In the next section we will see an example of this.
+
+
+## Test example and good practices
+In this section I want to wrap up and show a little example. This is the tiny reality: In the project we have a Django app called users where we have the user model that have a username, phone_number, and a category in the system, that could be admin, common user, and guest. The user can sign-up and login in the application. 
+
+### Model and factory definition
+
+```python
+# model.py in the Django app called users
+from django.db import models
+from django.contrib.auth.models import AbstractUser
+
+
+# class to define choices
+class Category(models.TextChoices):
+    GUEST = 'G', 'Guest user'
+    COMMON_USER = 'C', 'Common user'
+    ADMIN = 'A', 'Admin user'
+
+
+# user model that inherits from AbstractUser
+# there is no need to define a username field
+# because is defined in the parent class
+class User(AbstractUser):
+    phone_number = models.CharField(max_length=50, blank=True, null=True)
+    category = models.CharField(
+        max_length=1,
+        choices=Category.choices,
+        default='G',
+    )
+```
+
+Now let's see the implementation of the user factory. This file is inside a test folder in the users Django app:
+
+```python
+# factory.py inside users/test
+from faker import Faker as FakerClass
+from typing import Any, Sequence
+from factory import django, Faker, post_generation
+
+from users.models import User, Category
+
+
+CATEGORIES_VALUES = [x[0] for x in Category.choices]
+
+
+class UserFactory(django.DjangoModelFactory):
+
+    class Meta:
+        model = User
+    
+    username = Faker('user_name')
+    phone_number = Faker('phone_number')
+    category = Faker('random_element', elements=CATEGORIES_VALUES)
+
+    @post_generation
+    def password(self, create: bool, extracted: Sequence[Any], **kwargs):
+        password = (
+            extracted
+            if extracted
+            else FakerClass().password(
+                length=30,
+                special_chars=True,
+                digits=True,
+                upper_case=True,
+                lower_case=True,
+            )
+        )
+        self.set_password(password)
+
+```
+
+A couple of things about this:
+- In the `Meta` class inside the factory, we have to indicate the corresponding model.
+- In the category attribute, we use `random_element` of Faker to randomly select a choice of the Category choice list. We've defined `CATEGORIES_VALUES` to specify which part of the choice we want to take. That is because each element in the choices list is a tuple, where the first part is the value, and the second one the explanatory text. So that `CATEGORIES_VALUES` set has only that first part for each element.
```

Comment:
```suggestion
- In the category attribute, we use `random_element` of Faker to randomly select a choice of the Category choice list. We've defined `CATEGORIES_VALUES` to specify which part of the choice we want to take. That is because each element in the choices list is a tuple, where the first part is the value, and the second one the explanatory text. So, that `CATEGORIES_VALUES` set has only that first part for each element.
```